design/ceph-volume-provisioning.md

```diff
@@ -0,0 +1,122 @@
+- Partitioning and configuration of a `metadata` device where the WAL and DB are placed on a different device from the data
+- Support for both directories and devices
+- Support for bluestore and filestore
+
+Since this is mostly handled by `ceph-volume` now, Rook should replace its own provisioning code and rely on `ceph-volume`.
+
+## ceph-volume Design
+
+`ceph-volume` is a CLI tool included in the `ceph/ceph` image that will be used to configure and run Ceph OSDs.
+`ceph-volume` will replace the OSD provisioning mentioned previously in the legacy design.
+
+At a high level this flow remains unchanged from the flow in the [one-osd-per-pod design](dedicated-osd-pod.md#create-new-osds).
+No new jobs or pods need to be launched from what we have today. The sequence of events in the OSD provisioning will be the following.
+- The cluster CRD specifies what nodes/devices to configure with OSDs
+- The operator starts a provisioning job on each node where OSDs are to be configured
+- The provisioning job:
+  - Detects what devices should be configured
+  - Calls `ceph-volume lvm batch` to prepare the OSDs on the node. A single call is made with all of the devices unless more specific settings are included for LVM and partitions.
+  - Calls `ceph-volume lvm list` to retrieve the results of the OSD configuration. Store the results in a configmap for the operator to take the next step.
+- The operator starts a deployment for each OSD that was provisioned. `rook` is the entrypoint for the container.
+  - `ceph-volume` is called to `activate` the osd, which mounts the config directory such as `/var/lib/ceph/osd-0`, using a tempfs mount
+  - The OSD daemon is started with `ceph-osd`
+  - When `ceph-osd` exits, `rook` will exit and the pod will be restarted by K8s.
+
+### New Features
+
+`ceph-volume` enables rook to expose several new features:
+- Multiple OSDs for a single device, which is ideal for NVME devices.
+- Configure OSDs on LVM, either consuming the existing LVM or automatically configuring LVM on the raw devices.
+- Encrypt the OSD data with dmcrypt
+
+The Cluster CRD will be updated with the following settings to enable these features. All of these settings can be specified
+globally if under the `storage` element as in this example. The `config` element can also be specified under individual
+nodes or devices.
+```yaml
+  storage:
+    config:
+      # whether to encrypt the contents of the OSD with dmcrypt
+      encryptDevice: true
+      # how many OSDs should be configured on each device. only recommended to be greater than 1 for NVME devices
+      osdsPerDevice: 1
+      # the class name for the OSD(s) on devices
+      crushDeviceClass: ssd
+```
+
+If more flexibility is needed that consuming raw devices, LVM or partition names can also be used for specific nodes.
+
+```yaml
+  storage:
+    nodes:
+    - name: node2
+      # OSDs on LVM
+      logicalDevices:
+      # the DB, WAL, and Data are on separate LVs
+      - db: db_lv1
```

Comment:
These settings would be on the rook.io types, but they're not really applicable beyond ceph types. @jbw976 any thoughts on this? Using the `config` attributes under a device doesn't seem right to me since it's a different type from a raw device.

Reply:
@jbw976 I'm out of time to implement these options for 0.9. Are you ok if we merge the design doc with this open question and re-evaluate later? Per our conversation I still need to show what it looks like under the more general `config` settings. I added a comment to this effect above the `logicalDevices` element.